### ISSUE TITLE
device TCTI must read all data in one read

### ISSUE THREAD

**Issue by flihp** [tpm2-software/tpm2-tss#947]
The device TCTI currently makes a bad assumption about reading data from the device driver: that we can read the TPM2 response header in one `read` operation (the first 10 bytes) to get the size of the whole thing, and then read the rest of the response in a second read. This is a bad assumption since the driver will close the file descriptor if we don't read the full response in one `read` operation.

This makes supporting the TCTI `receive` function difficult so we're gonna have to come up with a work-around.

**Comment by martinezjavier** [comment]
@flihp is expected that the driver will close the file descriptor if the whole response isn't read in one `read` operation? Should the driver be fixed instead?

**Comment by AndreasFuchsSIT** [comment]
@martinezjavier This is the particular behavior of the TPM driver on Linux. The TPM itself allows the platform to wait after having read the header before reading the rest of the response.

I do know that several Kernel drivers behave this way, e.g. under the /sysfs.

I don't know enough about the Kernel as a whole whether this is a kernel convention or anything.
If you want to tackle the kernel driver, it could also use a poll() implementation... ;-)

**Comment by martinezjavier** [comment]
@AndreasFuchsSIT yeah, that's what I thought but wasn't sure if it was a constraint of the TPM. I may take a look to the driver but I probably won't have time to do it before next week.

In general I prefer us to avoid doing workarounds, following was another example of an issue we had and thought first about doing a workaround in tpm2-tss but but at the end fixed the kernel implementation:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=095531f891e627e408606f2da4008d3d53e6748a

**Comment by AndreasFuchsSIT** [comment]
@martinezjavier Sounds great...

Just a random thought: Does the TPM-device-driver support `stat()` ? That would also be a way to determine the response's size, whilst keeping the current semantic of one-time-reads.

**Comment by martinezjavier** [comment]
@AndreasFuchsSIT I don't think that the `struct stat .st_size` makes sense for a character device since isn't really a file. But I don't see why reading the available buffer in two `read` syscalls as @flihp implemented shouldn't be supported by a driver.

**Comment by flihp** [comment]
I haven't seen any disagreement over whether or not we'd like to see the kernel support these things. It's just a matter of having the time to spend on it.

**Comment by martinezjavier** [comment]
Agreed, and even if the driver is eventually fixed we need your patches so it works with kernel versions that don't implements the new behavior.
